### PR TITLE
Fix equality and hashing in Python 3

### PR DESCRIPTION
**Pull request by StephanErb**
The combination of `__ne__`, `__eq__`, and `__hash__` is surprisingly tricky especially if multiple Python versions are in play:

*  When defining `__eq__` in Python 2, one should also define `__ne__` so that the operators will behave as expected. This is no longer necessary in Python 3 as there is a default `__ne__` implementation that delegates to `__eq__` and inverts the result. 
* It is okay to implement `__ne__` in a base class and make it available to subclasses. This is done in pystachio for any class deriving from `base.Object`. Classes with a custom `__eq__` which are not deriving from `base.Object` still need a custom `__ne__` implementation to make Python 2 work as expected.
* Classes which inherit a `__hash__` method from a parent class but change the meaning of  `__eq__` behave differently in Python 2 and Python 3. In Python 2 one can silently break the contract that equal objects should also have the same hashcode. Python 3 prevents this issue. A class that overrides `__eq__` and does not define `__hash__` will have its `__hash__` implicitly set to None. 

In summary:
* We always have to reimplement `__hash__` when we reimplement `__eq__` for a class. We cannot rely on inheritance here.
* When implementing `__eq__` we also need to have an implementation of `__ne__`. It is okay to inherit this implementation from a parent class.

For further information see https://docs.python.org/3/reference/datamodel.html#object.__hash__ and  https://docs.python.org/2/reference/datamodel.html#object.__hash__

### CODE REVIEW

**Comment by StephanErb** [comment]
@shanmugh @jsirois as you have been the most recent contributors to pystachio, any chance you can take a look? Thanks!

**Comment by jsirois** [comment]
@StephanErb let me know what you think about the "missing" `__ne__`. After that I'll merge and prep a release.

**Comment by StephanErb** [comment]
@jsirois thanks for taking a look! I have extended the test coverage to include more combinations of the of the tricky behaviour of `==` and `!=`. I have also significantly extended the PR description to clarify the behaviour.

**Comment by jsirois** [comment]
Thanks for the fixes and the education @StephanErb!

**Comment by StephanErb** [comment]
Great, thanks for the quick turnaround! :)
